### PR TITLE
Huffman Encoding using C++ STL

### LINKED ISSUE #1505: Adding Huffman Coding using C++
I want to implement **Huffman Coding** with _C++ STL_. I will be using _priority_queue_, _vector_ and others if required.
Please assign me this issue for HactoberFest 2019. @jainaman224

### PR DESCRIPTION
**Pull request by SubhradeepSS**
Fixes #1505 

#### Checklist:
- [x] 4 space indentation.
- [x] Coding conventions are followed.
- [x] Input is taken dynamically.
- [x] Sample Input / Output is added at the end of file.
- [x] Logic Documentation (Comments).
- [x] File names are correct.

Make sure these boxes are checked before your pull request (PR) is ready to be reviewed and merged. Thanks!

#### Changes proposed in this pull request:
- Added the code for Huffman Encoding using C++ Standard Template Library(STL).
- 

#### Languages Used:
- C++
- 

#### Files Added:
- [Huffman_Encoding using STL.cpp](https://github.com/SubhradeepSS/Algo_Ds_Notes/blob/Huffman_Encoding/Huffman_Encoding/Huffman_Encoding%20using%20STL.cpp)
- 

> We're happy to help you get this ready -- don't be afraid to ask for help.

Thanks!

### CODE REVIEW

File: Huffman_Encoding/Huffman_Encoding using STL.cpp [hunk 1]
```diff
@@ -0,0 +1,94 @@
+#include<bits/stdc++.h>
+using namespace std;
+
+// A Huffman tree node 
+struct MinHeap_Node
+{
+  int count; // Count of the character
```
Comment (jainaman224) [inline Huffman_Encoding/Huffman_Encoding using STL.cpp]: Use 4 space indentation

File: Huffman_Encoding/Huffman_Encoding using STL.cpp [hunk 2]
```diff
@@ -0,0 +1,94 @@
+#include<bits/stdc++.h>
```
Comment (jainaman224) [inline Huffman_Encoding/Huffman_Encoding using STL.cpp]: Fix file name.

File: Huffman_Encoding/Huffman_Encoding using STL.cpp [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+  print_node (node->right, s + "1");
+}
+
+// The main function that builds a Huffman Tree and print codes by traversing the built Huffman Tree 
+void Huffman (int count[], char data[], int n)
+{  
+  priority_queue < MinHeap_Node *, vector < MinHeap_Node * >, compare > heap;    // Min heap & inserting all characters of data[] 
+  for (int i = 0; i < n; i++)
+    queue.push (new MinHeap_Node (count[i], data[i]));
+  while (heap.size () != 1)
```
Comment (jainaman224) [inline Huffman_Encoding/Huffman_Encoding using STL.cpp]: Fix indentation.

File: Huffman_Encoding/Huffman_Encoding using STL.cpp (hunk 2 above)
Comment (SubhradeepSS) [reply Huffman_Encoding/Huffman_Encoding using STL.cpp]: What should be the name?

**Comment by jainaman224** [comment]
@SubhradeepSS Are you working on it?

**Comment by SubhradeepSS** [comment]
> @SubhradeepSS Are you working on it?

@jainaman224 I asked a query about file name?You didn't reply.
Also a little doubt:4space indentation needs to be applied for which areas?comments or in code part also?

File: Huffman_Encoding/Huffman_Encoding using STL.cpp (hunk 2 above)
Comment (jainaman224) [reply Huffman_Encoding/Huffman_Encoding using STL.cpp]: Ohh sorry, I missed this.

Comment (jainaman224) [reply Huffman_Encoding/Huffman_Encoding using STL.cpp]: Huffman_Encoding_STL.cpp

Will be better.

**Comment by jainaman224** [comment]
4 space indentation needs to be applied where we use indentation.
It is applicable for code and comment both.

**Comment by SubhradeepSS** [comment]
@jainaman224 Please check whether the formatting is correct or not

File: Huffman_Encoding/Huffman_Encoding_STL.cpp [hunk 4]
```diff
@@ -0,0 +1,100 @@
+#include<bits/stdc++.h>
+using   namespace   std;
+
+//A    Huffman    tree    node 
+struct  MinHeap_Node
+{
+  int   count;//Count   of    the    character
```
Comment (somya-kapoor) [inline Huffman_Encoding/Huffman_Encoding_STL.cpp]: Comments should be above the statement.

File: Huffman_Encoding/Huffman_Encoding_STL.cpp [hunk 5]
```diff
@@ -0,0 +1,100 @@
+#include<bits/stdc++.h>
+using   namespace   std;
+
+//A    Huffman    tree    node 
+struct  MinHeap_Node
+{
+  int   count;//Count   of    the    character 
+  char  data; //One    of    the    input    characters 
+  //Left    and    right   child 
+  MinHeap_Node    *left,*right;
```
Comment (somya-kapoor) [inline Huffman_Encoding/Huffman_Encoding_STL.cpp]: Please correct the indentation

File: Huffman_Encoding/Huffman_Encoding_STL.cpp [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+    MinHeap_Node(int    count,char    data)
+  {
+    left=right=NULL;
```
Comment (somya-kapoor) [inline Huffman_Encoding/Huffman_Encoding_STL.cpp]: inconsistence in indentation

**Comment by somya-kapoor** [comment]
Please correct the indentation. Also, add space around the operators

**Comment by somya-kapoor** [comment]
@SubhradeepSS Are you working on it?

**Comment by SubhradeepSS** [comment]
> 
> 
> @SubhradeepSS Are you working on it?

Yes I am working,Actually was busy during the last few days as I had to come back home due to corona outbreak.

**Comment by SubhradeepSS** [comment]
@jainaman224 @somya-kapoor please check whether the format is ok or not?

**Comment by somya-kapoor** [comment]
Please correct the indentation. Also, add space around the operators

**Comment by SubhradeepSS** [comment]
> 
> 
> Please correct the indentation. Also, add space around the operators(+,->,>>,<<)

Can you be more specific about indentation?I am not getting about what exact is required for indentation?

File: Huffman_Encoding/Huffman_Encoding_STL.cpp [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+};
+
+//Prints    huffman    codes    from    the    root    of    Huffman    Tree
+void    print_node(MinHeap_Node    *node,string    s)
+{
+  if(!node)
+    return;
+  if(node->data != '#')
+    cout<<node->data<<"    "<<s<<endl;
+  print_node(node->left, s + "0");
```
Comment (somya-kapoor) [inline Huffman_Encoding/Huffman_Encoding_STL.cpp]: add space around ->

File: Huffman_Encoding/Huffman_Encoding_STL.cpp [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+  }
+};
+
+//Prints    huffman    codes    from    the    root    of    Huffman    Tree
+void    print_node(MinHeap_Node    *node,string    s)
+{
+  if(!node)
+    return;
+  if(node->data != '#')
+    cout<<node->data<<"    "<<s<<endl;
```
Comment (somya-kapoor) [inline Huffman_Encoding/Huffman_Encoding_STL.cpp]: add space around <<

File: Huffman_Encoding/Huffman_Encoding_STL.cpp [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+struct    compare    
+{
+  bool    operator()(MinHeap_Node    *l, MinHeap_Node    *r)
+  {
+    return l->count > r->count;
+  }
+};
+
+//Prints    huffman    codes    from    the    root    of    Huffman    Tree
+void    print_node(MinHeap_Node    *node,string    s)
```
Comment (somya-kapoor) [inline Huffman_Encoding/Huffman_Encoding_STL.cpp]: Add a space after ,

File: Huffman_Encoding/Huffman_Encoding_STL.cpp [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+{
+  if(!node)
```
Comment (somya-kapoor) [inline Huffman_Encoding/Huffman_Encoding_STL.cpp]: Use 4 space indentation

File: Huffman_Encoding/Huffman_Encoding_STL.cpp [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+    cout<<node->data<<"    "<<s<<endl;
+  print_node(node->left, s + "0");
+  print_node(node->right,s + "1");
+}
+
+//The    main    function    that    builds    a    Huffman    Tree 
+//and    print    codes    by    traversing    the    built    Huffman    Tree 
+void    Huffman(int    count[],char    data[],int    n)
+{  
+//Min    heap    &    inserting    all    characters    of    data[]
```
Comment (somya-kapoor) [inline Huffman_Encoding/Huffman_Encoding_STL.cpp]: inconsistency in indentation

File: Huffman_Encoding/Huffman_Encoding_STL.cpp [hunk 12]
```diff
@@ -0,0 +1,102 @@
...
+    }
+    //Print    Huffman    codes    using    the    Huffman    tree
+  print_node(heap.top(), "");
+}
+
+int main ()
+{
+  cout<<"Enter    the    number    of    input    characters:    ";
+  int    n;
+  cin>>n;
```
Comment (somya-kapoor) [inline Huffman_Encoding/Huffman_Encoding_STL.cpp]: add space around >>

**Comment by SubhradeepSS** [comment]
@somya-kapoor @jainaman224 Please review my PR

**Comment by somya-kapoor** [comment]
@SubhradeepSS Please use 4 space indentation. For reference visit any file.

**Comment by SubhradeepSS** [comment]
> 
> 
> @SubhradeepSS Please use 4 space indentation. For reference visit any file present in the repository.

Is the formatting proper now?

**Comment by SubhradeepSS** [comment]
@jainaman224 @somya-kapoor  Please review

**Comment by SubhradeepSS** [comment]
@somya-kapoor @jainaman224 Please review the PR

File: Huffman_Encoding/Huffman_Encoding_STL.cpp [hunk 13]
```diff
@@ -0,0 +1,100 @@
...
+    bool operator()(MinHeap_Node *l, MinHeap_Node *r){
+        return l -> count > r -> count;
+    }
+};
+
+//Prints huffman codes from the root of Huffman Tree
+void print_node(MinHeap_Node *node, string s){
+    if(!node)
+        return;
+    if(node->data != '#')
```
Comment (somya-kapoor) [inline Huffman_Encoding/Huffman_Encoding_STL.cpp:32]: Add space around ->

File: Huffman_Encoding/Huffman_Encoding_STL.cpp [hunk 14]
```diff
@@ -0,0 +1,100 @@
+#include<bits/stdc++.h>
+using namespace std;
+
+//A Huffman tree node
```
Comment (somya-kapoor) [inline Huffman_Encoding/Huffman_Encoding_STL.cpp:4]: Add space after //

File: Huffman_Encoding/Huffman_Encoding_STL.cpp [hunk 15]
```diff
@@ -0,0 +1,100 @@
+#include<bits/stdc++.h>
+using namespace std;
+
+//A Huffman tree node 
+struct MinHeap_Node{    
+    //Count of the character
```
Comment (somya-kapoor) [inline Huffman_Encoding/Huffman_Encoding_STL.cpp:6]: Add space after // . do this whole file

File: Huffman_Encoding/Huffman_Encoding_STL.cpp [hunk 16]
```diff
@@ -0,0 +1,100 @@
...
+    cout << "\nEnter the characters and corresponding count: ";
+    char data[n];
+    int count[n];
+    for(int i = 0; i < n; i++)
+        cin >> data[i] >> count[i];
+
+    cout << "Following is the Huffman Encoding:\n";
+    Huffman(count, data, n);
+}
+
```
Comment (somya-kapoor) [inline Huffman_Encoding/Huffman_Encoding_STL.cpp:82]: Remove this line

Comment (SubhradeepSS) [reply Huffman_Encoding/Huffman_Encoding_STL.cpp:81]: Capture
@somya-kapoor  I removed the 79th line as shown in screenshot but it is not showing in here.
Please check and review

Comment (SubhradeepSS) [reply Huffman_Encoding/Huffman_Encoding_STL.cpp:81]: @somya-kapoor Please check this also

Comment (somya-kapoor) [reply Huffman_Encoding/Huffman_Encoding_STL.cpp:82]: Line is not removed, remove 82

**Comment by somya-kapoor** [comment]
After making the changes, please squash the commits.

**Comment by SubhradeepSS** [comment]
> 
> 
> After making the changes, please squash the commits.

@somya-kapoor I have squashed the commits

**Comment by SubhradeepSS** [comment]
@somya-kapoor Can u please reopen my PR #1956 which was closed.I have made the changes and pushed.Can u please review the PR.The PR was for issue #1478 which was assigned to me in c++.

**Comment by SubhradeepSS** [comment]
@somya-kapoor Please review my PR.Squashing of commits is done.

**Comment by SubhradeepSS** [comment]
@somya-kapoor Please review my closed PR #1956 for the issue #1478 which was assigned to me.The pr was closed without notifying me.Please open the pr and review it.
Capture

**Comment by SubhradeepSS** [comment]
@somya-kapoor Is anything left to be done for the PR?

**Comment by SubhradeepSS** [comment]
> 
> 
> Squash the commits
@somya-kapoor 
The extra commits are those of updating with the original repo.I already squashed the commits associated with my work earlier.Do I need to squash the update commits also?

**Comment by somya-kapoor** [comment]
> > Squash the commits
> > @somya-kapoor
> > The extra commits are those of updating with the original repo.I already squashed the commits associated with my work earlier.Do I need to squash the update commits also?

Yes

**Comment by SubhradeepSS** [comment]
> > > Squash the commits
> > > @somya-kapoor
> > > The extra commits are those of updating with the original repo.I already squashed the commits associated with my work earlier.Do I need to squash the update commits also?
> 
> Yes

Done

**Comment by SubhradeepSS** [comment]
@somya-kapoor Please review

**Comment by SubhradeepSS** [comment]
@somya-kapoor Please review my PR

**Comment by SubhradeepSS** [comment]
@somya-kapoor @jainaman224 Please review the PR and tell what to update?

**Comment by somya-kapoor** [comment]
@SubhradeepSS Please delete the unwanted files, there are around 180 files in this PR

**Comment by SubhradeepSS** [comment]
@

> 
> 
> @SubhradeepSS Please delete the unwanted files, there are around 180 files in this PR

@somya-kapoor I have updated my branch with the original repo.While rebasing my last two commits,all the 168 commits done by others appear in the terminal.
For now,2commits are there for me and files have been updated correctly.
Please allow the 2commits for this case as changing all the 168 files is quite hard.

**Comment by somya-kapoor** [comment]
@SubhradeepSS Did you ran your code because I am getting an error while running it.

**Comment by SubhradeepSS** [comment]
> 
> 
> @SubhradeepSS Did you ran your code because I am getting an error while running it.

@somya-kapoor Updated the code

**Comment by SubhradeepSS** [comment]
@somya-kapoor review

**Comment by SubhradeepSS** [comment]
@somya-kapoor @singh-shreya6 Please review this PR also

**Comment by SubhradeepSS** [comment]
@singh-shreya6 @somya-kapoor  Please review my PR

**Comment by SubhradeepSS** [comment]
@somya-kapoor @singh-shreya6 Please review the PR

**Pull request merged**